js/widgets/behaviors/graph-camera.ts

```diff
@@ -0,0 +1,270 @@
+      new THREE.Matrix4().multiplyMatrices(
+        camera.projectionMatrix,
+        camera.matrixWorldInverse
+      )
+    );
+
+    for (let node of graph.graphData().nodes) {
+      let __threeObj: THREE.Object3D = (node as any).__threeObj;
+      if (
+        frustum.containsPoint(__threeObj.position) ||
+        frustum.intersectsObject(__threeObj)
+      ) {
+        visible.push(i);
+      }
+      i++;
+    }
+
+    return visible;
+  }
+
+  protected getVisible2d(zoom: IZoomData): number[] {
+    const { graph, k } = zoom;
+    const halfW = graph.width() / 2 / k;
+    const halfH = graph.height() / 2 / k;
+    const bx = [zoom.x - halfW, zoom.x + halfW];
+    const by = [zoom.y - halfH, zoom.y + halfH];
+    const visible = [];
+    let i = 0;
+    for (let { x, y } of graph.graphData().nodes) {
+      if (x >= bx[0] && x <= bx[1] && y >= by[0] && y <= by[1]) {
+        visible.push(i);
+      }
+      i++;
+    }
+    return visible;
+  }
+}
+
+export class GraphDirectorModel extends ZoomBase implements IBehave {
+  static model_name = 'GraphDirectorModel';
+
+  private _resetting = false;
+
+  static serializers = {
+    ...FacetedModel.serializers,
+    visible: widget_serialization,
+  };
+
+  protected get _modelClass(): typeof GraphDirectorModel {
+    return GraphDirectorModel;
+  }
+
+  defaults() {
+    return {
+      ...super.defaults(),
+      ...WIDGET_DEFAULTS,
+      _model_name: GraphCameraModel.model_name,
+      zoom: null,
+      visible: null,
+      padding: null,
+      center: null,
+      zoom_first: false,
+      zoom_duration: 0.2,
+      pan_duration: 0.2,
+      fit_duration: 0.2,
+    };
+  }
+
+  initialize(attributes: Backbone.ObjectHash, options: IBackboneModelOptions) {
+    super.initialize(attributes, options);
+    this.on(
+      'change:zoom change:center change:visible change:padding',
+      this.onZoomChanged,
+      this
+    );
+    this.updateRequested.connect(this.onZoomChanged, this);
+    this.onZoomChanged();
+  }
+
+  get zoomFirst(): boolean {
+    return this.get('zoom_first');
+  }
+
+  set zoomFirst(zoomFirst: boolean) {
+    this.set('zoom_first', zoomFirst);
+  }
+
+  get zoomDuration(): number {
+    return (this.get('zoom_duration') || 0) * 1000;
+  }
+
+  set zoomDuration(zoomDuration: number) {
+    this.set('zoom_duration', zoomDuration);
+  }
+
+  get panDuration(): number {
+    return (this.get('pan_duration') || 0) * 1000;
+  }
+
+  set panDuration(panDuration: number) {
+    this.set('pan_duration', panDuration);
+  }
+
+  get fitDuration(): number {
+    return (this.get('fit_duration') || 0) * 1000;
+  }
+
+  set fitDuration(fitDuration: number) {
+    this.set('fit_duration', fitDuration);
+  }
+
+  get fitPadding(): number {
+    return this.get('fit_padding');
+  }
+
+  set fitPadding(fitPadding: number) {
+    this.set('fit_padding', fitPadding);
+  }
+
+  onZoomChanged() {
+    if (!this._resetting) {
+      this._graphCameraUpdateRequested.emit();
```

Comment:
I'm getting an error here some times:
![image](https://github.com/jupyrdf/ipyforcegraph/assets/1438114/3c5111ca-f089-43cc-901c-0e3c9009ca17)

Not sure what is causing it, but it happens when I play with the `look at` controls in 3D in the `Camera` notebook.

Reply:
Bad bounds checking: 050b0a1